src/include/ikarus/controlRoutines/PathFollowingTechnique.hh

```diff
@@ -0,0 +1,94 @@
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#pragma once
+#include "src/include/ikarus/utils/PathFollowingFunctions.hh"
+
+#include <memory>
+
+#include <Eigen/Core>
+
+#include <ikarus/linearAlgebra/nonLinearOperator.hh>
+#include <ikarus/solver/nonLinearSolver/newtonRaphsonWithScalarSubsidiaryFunction.hh>
+#include <ikarus/utils/observer/observer.hh>
+#include <ikarus/utils/observer/observerMessages.hh>
+
+namespace Ikarus {
+
+  struct PathFollowingInformation {
+    bool success{false};
+  };
+
+  template <typename NonLinearSolver, typename PathFollowingType = Ikarus::StandardArcLength>
+  requires Concepts::PathFollowingStrategy<PathFollowingType, typename NonLinearSolver::NonLinearOperator>
+  class PathFollowing : public IObservable<ControlMessages> {
+  public:
+    PathFollowing(const std::shared_ptr<NonLinearSolver>& p_nonLinearSolver, int loadSteps, double stepSize,
+                  PathFollowingType p_pathFollowingType = Ikarus::StandardArcLength{})
+        : nonLinearSolver{p_nonLinearSolver},
+          loadSteps_{loadSteps},
+          stepSize_{stepSize},
+          pathFollowingType_{p_pathFollowingType} {}
+
+    PathFollowingInformation run() {
+      PathFollowingInformation info({false});
+      auto& nonOp = nonLinearSolver->nonLinearOperator();
+      this->notify(ControlMessages::CONTROL_STARTED);
+
+      SubsidiaryArgs subsidiaryArgs;
+
+      subsidiaryArgs.stepSize = stepSize_;
+      subsidiaryArgs.DD.resizeLike(nonOp.firstParameter());
+      subsidiaryArgs.DD.setZero();
+
+      /// Initializing solver
+      this->notify(ControlMessages::STEP_STARTED);
+      auto subsidiaryFunctionPtr = [&](auto&& args) { return pathFollowingType_.evaluateSubsidiaryFunction(args); };
+      auto solverInfo            = nonLinearSolver->solve(subsidiaryFunctionPtr, subsidiaryArgs);
+      if (not solverInfo.success) return info;
+      this->notify(ControlMessages::SOLUTION_CHANGED);
+      this->notify(ControlMessages::STEP_ENDED);
+
+      /// Calculate predictor for a particular step
+      for (int ls = 0; ls < loadSteps_; ++ls) {
+        this->notify(ControlMessages::STEP_STARTED);
+
+        if (ls > 0)
+          pathFollowingType_.intermediatePrediction(nonOp, subsidiaryArgs);
+        else
+          pathFollowingType_.initialPrediction(nonOp, subsidiaryArgs);
```

Comment:
Can this be moved before the loop and first be evaluated?
Then psi is first defined